### DISCUSSION TITLE
Does GoodJob work with ActiveSupport::CurrentAttributes?

### DISCUSSION THREAD

**Discussion by JonathanFrias** [bensheldon/good_job#1260] [Q&A]
If I set a `Current.user` for example, does it get set properly, or do I have to manually manage this? I'm seeing some errors around this class, and I'm wondering if I'm doing something that's not supported.

**Comment by bensheldon** [comment]
You have to manually manage it. I'd accept an Active Job extension if you wanted to propose one. I discuss it a little here: https://github.com/bensheldon/good_job/discussions/901#discussioncomment-5394194

It wouldn't really have any dependency on GoodJob, but just serialized the CurrentAttributes on enqueue, and deserialize them on perform. Should probably pass them through ActiveJob serializer too so that it works with GlobalID.

I do though think the interface should be explicit like:

```ruby
Current.user = User.find(session[:current_user_id])
MyJob.perform_later

class MyJob < ApplicationJob
  maintain_current_attributes Current: [:user, :something_else] # I dunno what the best way is to list multiple constants and their properties

  def perform
    Current.user # => would work
  end
end

**Reply by JonathanFrias** [reply]
Oh that interesting, I didn't design my jobs like that. I pass the current attributes to the job and then set them again in the job context manually. I'm wondering if an example like the following would have any kind of race condition between the jobs/threads/processes. 

```ruby

def perform(user, post)
  Current.user = user
  post.comments.each do |post|
    Current.user = post.user
    # do something as that user
  end
end
```

**Reply by bensheldon** [reply]
oh! I misunderstood your question. 

What you're doing should work fine. GoodJob runs each job in its own thread. Even if you're using `inline` execution mode, Active Job wraps all `#perform` with a Rails Executor that will reset all the CurrentAttributes after your job.
